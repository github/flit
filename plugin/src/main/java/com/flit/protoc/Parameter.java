package com.flit.protoc;

import com.flit.protoc.gen.GeneratorException;
import java.util.Arrays;
import java.util.Map;
import java.util.function.Function;
import java.util.stream.Collectors;
import lombok.Getter;
import lombok.ToString;

@Getter @ToString public class Parameter {

  public static final String PARAM_TARGET = "target";
  public static final String PARAM_CLIENT = "client";
  public static final String PARAM_TYPE = "type";
  public static final String PARAM_CONTEXT = "context";
  public static final String PARAM_REQUEST = "request";

  private final String key;
  private final String value;

  public Parameter(String[] strings) {
    this.key = strings[0];
    this.value = strings[1];
  }

  public static Map<String, Parameter> of(String value) {
    if (value == null || (value = value.trim()).isEmpty()) {
      throw new GeneratorException("Empty value passed to parameter builder");
    }

    return Arrays.stream(value.split(","))
      .map(p -> p.split("=", 2))
      .filter(s -> s.length == 2 && !s[0].trim().isEmpty() && !s[1].trim().isEmpty())
      .map(Parameter::new)
      .collect(Collectors.toMap(Parameter::getKey, Function.identity()));
  }
}
