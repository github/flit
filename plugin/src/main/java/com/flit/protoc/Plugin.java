package com.flit.protoc;

import static com.flit.protoc.Parameter.PARAM_REQUEST;
import static com.flit.protoc.Parameter.PARAM_TARGET;
import static com.flit.protoc.Parameter.PARAM_TYPE;

import com.flit.protoc.gen.Generator;
import com.flit.protoc.gen.GeneratorException;
import com.flit.protoc.gen.server.jaxrs.JaxrsGenerator;
import com.flit.protoc.gen.server.spring.SpringGenerator;
import com.flit.protoc.gen.server.undertow.UndertowGenerator;
import com.google.protobuf.compiler.PluginProtos.CodeGeneratorRequest;
import com.google.protobuf.compiler.PluginProtos.CodeGeneratorResponse;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.Map;

public class Plugin {

  private final CodeGeneratorRequest request;

  public Plugin(CodeGeneratorRequest request) {
    this.request = request;
  }

  public CodeGeneratorResponse process() {
    if (!request.hasParameter()) {
      return CodeGeneratorResponse.newBuilder()
          .setError("Usage: --flit_out=target=server,type=[spring|undertow|jaxrs][,request=[class(es)]]:<PATH>")
          .build();
    }

    Map<String, Parameter> params = Parameter.of(request.getParameter());
    try {
      CodeGeneratorResponse.Builder builder = CodeGeneratorResponse.newBuilder();
      resolveGenerator(params).generate(request, params).forEach(builder::addFile);
      return builder.build();
    } catch (GeneratorException e) {
      return CodeGeneratorResponse.newBuilder().setError(e.getMessage()).build();
    }
  }

  private Generator resolveGenerator(Map<String, Parameter> params) {
    if (!params.containsKey(PARAM_TARGET)) {
      throw new GeneratorException("No argument specified for target");
    }
    if (!params.containsKey(PARAM_TYPE)) {
      throw new GeneratorException("No argument specified for type");
    }
    List<String> requestServices = getRequestServices(params);
    switch (params.get(PARAM_TARGET).getValue()) {
      case "server":
        switch (params.get(PARAM_TYPE).getValue()) {
          case "boot":
          case "spring":
            return new SpringGenerator(requestServices);
          case "undertow":
            return new UndertowGenerator(requestServices);
          case "jaxrs":
            return new JaxrsGenerator(requestServices);
          default:
            throw new GeneratorException("Unknown server type: " + params.get(PARAM_TYPE).getValue());
        }
      default:
        throw new GeneratorException("Unknown target type: " + params.get(PARAM_TARGET).getValue());
    }
  }

  private List<String> getRequestServices(Map<String, Parameter> params) {
    Parameter requestServices = params.get(PARAM_REQUEST);
    if (requestServices == null) {
      return Collections.emptyList();
    } else {
      return Arrays.asList(requestServices.getValue().split(","));
    }
  }
}
