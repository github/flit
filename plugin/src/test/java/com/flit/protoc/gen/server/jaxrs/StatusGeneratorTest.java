package com.flit.protoc.gen.server.jaxrs;

import static java.util.stream.Collectors.toList;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;

import com.flit.protoc.Plugin;
import com.flit.protoc.gen.BaseGeneratorTest;
import com.google.protobuf.compiler.PluginProtos;
import com.google.protobuf.compiler.PluginProtos.CodeGeneratorResponse.File;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.Test;

/**
 * Tests the generation of a service that has core definition imported from another file
 */
public class StatusGeneratorTest extends BaseGeneratorTest {

  @Test
  public void test_Generate() throws Exception {
    PluginProtos.CodeGeneratorRequest request = loadJson("status.jaxrs.json");

    Plugin plugin = new Plugin(request);
    PluginProtos.CodeGeneratorResponse response = plugin.process();

    assertNotNull(response);
    assertEquals(2, response.getFileCount());

    assertEquals(response.getFile(0).getName(), "com/example/helloworld/RpcStatus.java");
    assertEquals(response.getFile(1).getName(), "com/example/helloworld/RpcStatusResource.java");

    Approvals.verifyAll("", response.getFileList().stream().map(File::getContent).collect(toList()));
    response.getFileList().forEach(BaseGeneratorTest::assertParses);
  }

}
