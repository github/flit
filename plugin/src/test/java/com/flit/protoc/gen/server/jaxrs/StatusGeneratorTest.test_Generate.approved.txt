[0] = package com.example.helloworld;

public interface RpcStatus {
  StatusOuterClass.StatusResponse handleGetStatus(Core.Empty in);
}

[1] = package com.example.helloworld;

import com.google.protobuf.util.JsonFormat;
import java.io.InputStreamReader;
import java.lang.Exception;
import java.nio.charset.StandardCharsets;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import javax.ws.rs.POST;
import javax.ws.rs.Path;
import javax.ws.rs.core.Context;

@Path("/twirp/com.example.helloworld.Status")
public class RpcStatusResource {
  private final RpcStatus service;

  public RpcStatusResource(RpcStatus service) {
    this.service = service;
  }

  @POST
  @Path("/GetStatus")
  public void handleGetStatus(@Context HttpServletRequest request,
      @Context HttpServletResponse response) throws Exception {
    boolean json = false;
    final Core.Empty data;
    if (request.getContentType().equals("application/protobuf")) {
      data = Core.Empty.parseFrom(request.getInputStream());
    } else if (request.getContentType().startsWith("application/json")) {
      json = true;
      Core.Empty.Builder builder = Core.Empty.newBuilder();
      try (InputStreamReader reader = new InputStreamReader(request.getInputStream(), StandardCharsets.UTF_8)) {
        JsonFormat.parser().merge(reader, builder);
      }
      data = builder.build();
    } else {
      response.setStatus(415);
      response.flushBuffer();
      return;
    }
    StatusOuterClass.StatusResponse retval = service.handleGetStatus(data);
    response.setStatus(200);
    if (json) {
      response.setContentType("application/json; charset=utf-8");
      response.getOutputStream().write(JsonFormat.printer().omittingInsignificantWhitespace().print(retval).getBytes(StandardCharsets.UTF_8));
    } else {
      response.setContentType("application/protobuf");
      retval.writeTo(response.getOutputStream());
    }
    response.flushBuffer();
  }
}

